Specs v2 and v3

cyberbikepunk opened the pull request:
cc @akariv @annafralberts @seegerseeger

cyberbikepunk committed:
Add raw files for berlin.

cyberbikepunk committed:
Resolve merge conflict on .gitignore.

cyberbikepunk committed:
Sandbox the bootstrapping module (part 1).

cyberbikepunk committed:
Sandbox the boostrap module (part 1).

cyberbikepunk committed:
Move the contents of data/nl and data/pl to the new folders.

cyberbikepunk committed:
Merge branch 'berlin' into feature/1093/bootstrap-the-datapackage-pipeline

cyberbikepunk committed:
Move Berlin data files to new Berlin folder.

cyberbikepunk committed:
Sandbox the bootstrap module (part 2).

- fix folder and filename bugs
- take care of corner case: invalid YAML

cyberbikepunk committed:
Sandbox the bootstrap module (part 3).

- rename update.py -> bootstrap.py
- change object imports to module imports to improve readability
- refactor some of the functions
- add a docstring to the module
- explicitly specify that the module on;y supports python3

cyberbikepunk committed:
Use a proper python .gitignore template and remove pycache.

cyberbikepunk committed:
Modify source folder files.

This is simply a consequence of running the modified bootstrap script.

cyberbikepunk committed:
Replace ValidationError with ValueError.

cyberbikepunk committed:
Improve the docstring.

cyberbikepunk committed:
Remove pycache.

Commit:
Malta 2007-2013 source description

cyberbikepunk committed:
Improve the code style and code logic.

- naming: file > stream
- YAML: preferred over JSON because it supports comments
- pipeline not triggered if YAML parser error is encountered
- changed % > format (except for logging methods)
- geocode never none: fixed

On multiples raises and conditional returns: I agree! I'm only leaving it only because this will soon disapear with source.description.v2.yaml.

cyberbikepunk committed:
Add source.description.v2.yaml

cyberbikepunk committed:
Add source.description.v2.yaml

cyberbikepunk commented:
@pwalsh got a 404?. I thought i was cleaner to branch this off master.

cyberbikepunk committed:
Merge pull request #105 from os-data/feature/1093/bootstrap-the-datapackage-pipeline

Feature/1093/bootstrap the datapackage pipeline

cyberbikepunk committed:
Merge pull request #102 from os-data/72-malta-sourcing

Malta 2007-2013 source description

cyberbikepunk commented:
Hey @akariv @pwalsh @pudo @seegerseeger @annafralberts There are things to take into consideration for getting this right. Here are my thoughts...

## File format:

YAML has been preferred over JSON because it supports comments and the entire things is written like a tutorial. However, processing the hand written files that are trickling down shows that @pwalsh's humble opinion about whitespace is not far from the truth. the question is kinda: do you wanna spend your time answering questions about how to fill the doc or do you prefer fixing white-space errors? I haven't given up on YAML yet because at this stage I haven't hammered that point to the contributors and maybe the situation can improve. @akariv 's suggestion of using an online web-app is promising.

## Version 1

So far we've been working with a much simpler file, where the contributor could ad-lib on the fields (s)he found on the data. The idea is that the user suggested the mapping by providing a translation and a  description of the field. I got very little of the latter, which is really what is supposed to help out the person writing the transform code downstream. If that person has to do the research again to find out how the fields map, we're losing.

## version 2

now we're basically suggesting all the columns directly inside the template. this makes it much easier for the user to fill in. The drawback is that  we're expecting a 1-to-1 mapping between columns in the raw data and those in the final fiscal package. I deliberately ignored corner cases but this probably was a mistake. e.g. I but hadn't anticipated a case like Austria:

```
        - name: Status G = genehmigt, A = ausbezahlt
          translation: G = commited, A = paid out
          type: string
          missingValue:
          description:
```

In  the above example, the we're expecting the columns `committed` and `paid out` and instead we're getting a single "measure" column binded to a "category" column. This is like trying to shove a cube through a round hole.

## Why not a datapackage format directly?

I was really putting myself in the shoes of the contributor here. This extra-step makes it more accessible and less error prone.
1. slugify the title to get the name (not that important)
2. supporting multiple identical resources with the same schema to avoid unmanageably long files (more-useful: there are many cases of data sets broken down into multiple files (error prone in my -humble- opinion)... maybe it would be nice if the datapackage specs could support that?.

## conclusion

 i'm thinking that the looser approach of version 1, combined with emphasis on contributor training on how to do proper field descriptions, combined with an optional web app like [this](http://exavolt.github.io/onde/) could be the winner.

What do you think? I would appreciate feedback in the meantime, I'm moving on to modeling the fiscal package now and pushing some of the datasets up to os.

pwalsh commented:
Hi!

My opinion about YAML and non-technies is not humble - I was just being polite :). 

YAML is a horrible format for _writing_ when the writer does not understand simple concepts a programmer takes for granted - like 3 spaces before a `-` actually having a completely different meaning to 2 spaces.

@cyberbikepunk I'm in complete agreement with @akariv that a web form is a much better solution, and if we were brought in to the conversation earlier, we could have helped shape this decision.

However, due to us having time constraints here, we need to take a wholistic view - will we lose momentum by shifting midstream? Is it quicker, for the current sourcing exercise, to just keep going and fix errors? My suggestion is that when @akariv gets back to work on Monday that you consult with him on this.

cyberbikepunk commented:
@pwalsh @akariv I'm thinking we should pursue as is. I can live with it. I just need to create the underlying schema and suggest the interface as an option. Although an interface without the possibility of adding guidelines is not ideal.

cyberbikepunk committed:
Run pipeline bootstrap on existing source files.

cyberbikepunk committed:
Merge pull request #109 from os-data/pipeline/data-processing

Run pipeline bootstrap on existing source description files.

cyberbikepunk committed:
Allow multiple resources.

cyberbikepunk committed:
Add source description file v3.

v1 will keep being supported. v2 has been dropped.

cyberbikepunk committed:
Copy official.fields.yaml to fiscal_field.yaml.

cyberbikepunk committed:
Merge branch 'task/1091/pipeline-specs' of github.com:os-data/eu-structural-funds into task/1091/pipeline-specs

cyberbikepunk committed:
Small fixes to source.description.v3.yaml.

cyberbikepunk committed:
Add fiscal.fields.yaml file.

The file "official.fields.yaml" is a misnomer: it really is the fiscal schema.  This is the schema common to all fiscal packages.

cyberbikepunk commented:
@akariv basically concerning `source.description.yaml`:
- v1 has been used up to now: will be support to avoid having to re-write everything
- v2 was a bad idea
- v3 is a better version of v1 and should be used from now on.

Commit:
fixed small typo

akariv commented on data/AT.austria/AT11.burgenland/pipeline-specs.yaml:
```diff
@@ -0,0 +1,3 @@
+AT11.burgenland:
+  pipeline: {params: source.datapackage.json, run: validate_datapackage}
+  schedule: {crontab: '@annually'}
```
this syntax won't work - you need the traditional crontab format

akariv commented on data/AT.austria/AT11.burgenland/pipeline-specs.yaml:
```diff
@@ -0,0 +1,3 @@
+AT11.burgenland:
```
Why are pipeline-specs in git anyway? Aren't they supposed to be generated automatically?

akariv commented on data/PL.poland/validate_tabular_remote_source.py:
```diff
@@ -0,0 +1,71 @@
+""" A processor to download a data-file from a link and validate it."""
```
I guess that this is still a work in progress?

What is the purpose of this code?

cyberbikepunk replied:
junk
